trin-execution/src/subcommands/era2.rs

```diff
@@ -0,0 +1,346 @@
+use std::sync::Arc;
+
+use alloy_rlp::{Decodable, EMPTY_STRING_CODE};
+use anyhow::{ensure, Error};
+use e2store::era2::{
+    AccountEntry, AccountOrStorageEntry, Era2, StorageEntry, StorageItem, MAX_STORAGE_ITEMS,
+};
+use eth_trie::{
+    decode_node,
+    node::{LeafNode, Node},
+    EthTrie, Trie, DB,
+};
+use ethportal_api::Header;
+use parking_lot::Mutex;
+use revm_primitives::{keccak256, B256, KECCAK_EMPTY, U256};
+use tracing::info;
+
+use crate::{
+    cli::{ExportState, ImportState},
+    execution::TrinExecution,
+    storage::{account::Account, account_db::AccountDB},
+    utils::full_nibble_path_to_address_hash,
+};
+
+#[derive(Debug)]
+struct LeafNodeWithKeyHash {
+    key_hash: B256,
+    value: Vec<u8>,
+}
+
+#[derive(Debug)]
+struct TrieNode {
+    node_hash: B256,
+    path: Vec<u8>,
+}
+
+impl TrieNode {
+    fn new(node_hash: B256, path: Vec<u8>) -> Self {
+        Self { node_hash, path }
+    }
+}
+
+struct TrieLeafIterator<TrieDB: DB> {
+    trie: Arc<Mutex<EthTrie<TrieDB>>>,
+    stack: Vec<TrieNode>,
+    leafs_to_process: Vec<LeafNodeWithKeyHash>,
+}
+
+impl<TrieDB: DB> TrieLeafIterator<TrieDB> {
+    pub fn new(trie: Arc<Mutex<EthTrie<TrieDB>>>) -> anyhow::Result<Self> {
+        let stack = vec![TrieNode::new(trie.lock().root_hash()?, vec![])];
+        Ok(Self {
+            trie,
+            stack,
+            leafs_to_process: vec![],
+        })
+    }
+
+    fn process_leaf(&mut self, leaf: Arc<LeafNode>, path: Vec<u8>) {
+        // reconstruct the address hash from the path so that we can fetch the
+        // address from the database
+        let mut partial_key_path = leaf.key.get_data().to_vec();
+        partial_key_path.pop();
+        let full_key_path = [&path, partial_key_path.as_slice()].concat();
+        let address_hash = full_nibble_path_to_address_hash(&full_key_path);
+        self.leafs_to_process.push(LeafNodeWithKeyHash {
+            key_hash: address_hash,
+            value: leaf.value.clone(),
+        });
+    }
+
+    fn process_node(&mut self, node: Node, path: Vec<u8>) -> anyhow::Result<()> {
```

Comment:
I think that era2 files must be canonical (meaning, there is only one valid version of it). To achieve that, all accounts should be sorted by account_hash (and similarly, storage entries should be sorted by key_hash).

This code doesn't do that. For example, Let's say this is the trie (let's assume with degree of 4):

```
B-Branch, L-Leaf, N-NULL
                           B_root
      /                /             \              \
     B1               L2              B3             N
  / / \ \                          / / \  \
L11 N  N N                        N N  L33 N
```
Desired order of leafs is `L11, L2, L33`.

If I'm not mistaken, after processing `B_root`, we would have these values:

```
stack = [B1, B3]
leafs_to_process = [L2]
```

And that would result in order of leafs being process: `L2, L33, L11`

Reply:
https://github.com/ethereum/trin/pull/1344#issuecomment-2338336517

Replying to both. I do agree this is an oversight in my implementation well it is small and can be fixed. I do agree the EthTrie iter should be used here.

At first look it traverses in order, one thing I am not sure of is if it traverses branch values properly which could be possible in the case of the storage trie.

A major complaint I have of the iter, is it hides error's

https://docs.rs/eth_trie/0.4.0/src/eth_trie/trie.rs.html#180 I think here an error should be returned. Because if this was to happen clearly a critical error has happened and the .era2 export would be corrupt so I will make a pr on ethereum/eth-trie.rs and ping you for review once that problem is resolved I will update this PR and we can continue the review from there

Reply:
If the branch's partial nimble of the path is 0x10 which is 16, shouldn't that be read after all the branch's children? But this iter reads the value first. 

Overall this iter seems todo anything correct, so if you have an idea if branch.value should be read first or last, I think everything else is good

Reply:
> If the branch's partial nimble of the path is 0x10 which is 16, shouldn't that be read after all the branch's children? But this iter reads the value first.

Now sure what you mean by this. Can you point me to the code line?

> Overall this iter seems todo anything correct, so if you have an idea if branch.value should be read first or last, I think everything else is good

While MPT allows for values to be stored in the branch trie node, Ethereum state trie doesn't use that feature, so I think it doesn't matter much. But I would say it makes sense to go first (which is what I think EthTrie does already, but I didn't look into details).

> https://docs.rs/eth_trie/0.4.0/src/eth_trie/trie.rs.html#180 I think here an error should be returned. Because if this was to happen clearly a critical error has happened and the .era2 export would be corrupt so I will make a pr on ethereum/eth-trie.rs and ping you for review once that problem is resolved I will update this PR and we can continue the review from there

Yes, I agree that this critical error shouldn't be ignored. I think it's better to make it iterate over `Result<_, TrieError>`, but panicking might be good enough as well.

Reply:
> > If the branch's partial nimble of the path is 0x10 which is 16, shouldn't that be read after all the branch's children? But this iter reads the value first.
> 
> Now sure what you mean by this. Can you point me to the code line?

```
impl TraceNode {
    fn advance(&mut self) {
        self.status = match &self.status {
            TraceStatus::Start => TraceStatus::Doing,
            TraceStatus::Doing => match self.node {
                Node::Branch(_) => TraceStatus::Child(0),
                _ => TraceStatus::End,
            },
            TraceStatus::Child(i) if *i < 15 => TraceStatus::Child(i + 1),
            _ => TraceStatus::End,
        }
    }
}
```

Anyways so for a branch node as I was saying before it will read the `value` first

Reply:
Yes, I think this is correct.

But this is irrelevant for Ethereum State because keys are always 32 bytes (64 nibbles) long, so they are always stored in leaf nodes (values in branch nodes will be empty).